packages/vite/src/node/plugins/modulePreloadPolyfill.ts

```diff
@@ -89,12 +89,12 @@ function polyfill() {
     return fetchOpts
   }
 
+  const fetchCache = {} as any
   function processPreload(link: any) {
-    if (link.ep)
-      // ep marker = processed
-      return
+    if (link.ep) return
     link.ep = true
-    // prepopulate the load record
+    if (fetchCache[link.href]) return
+    fetchCache[link.href] = true
```

Comment:
IIUC this part only makes sense with es-module-shims.
es-module-shims uses `fetch` to fetch the module and executes them with `import(blobUrl)`.
https://github.com/guybedford/es-module-shims/blob/e8eb942fdd65de8669184a4c1d8ad7dde21ea576/src/es-module-shims.js#L297
https://github.com/guybedford/es-module-shims/blob/e8eb942fdd65de8669184a4c1d8ad7dde21ea576/src/es-module-shims.js#L233

Because we don't fetch modules on our side and rely on the browsers mechanism, `fetchCache` won't be used when the modules are fetched.
So this `fetchCache[link.href]` part does the same thing with `link.ep`.

Reply:
Oh, I missed this one. I didn't even look for the cache as I assumed it was local because I thought module preload polyfill should work with regular imports. Is your understanding of the polyfill that we can't extract the module preload and we should use the whole thing? The polyfill isn't small and IMO we shouldn't mess with the way modules are imported to support module preload. Not by default, and users could add this polyfill themselves if they want that.

But I think `fetchCache` was added to fix an edge case, and it was working in the general case without it (even if there is a double request).

So, we should close this PR. Then find out if the polyfill is really working. I need some help evaluating this, I thought it was working but I'm not sure anymore now.

Maybe for 4.2 we should get back to the original use of `'preload'` in browsers that don't support `'modulepreload'`. See https://github.com/vitejs/vite/commit/cb75dbd4a2b1c56e2a7aeaf1625818a4d1865f00#diff-121079017d1fa98d6a0ea3354d5c73df45ab277078228cd7384c7a4e84c5a813L35.

Reply:
Yeah I agree too. Looks like `fetchCache` doesn't have a benefit here. I think reverting to using `preload` would be the best bet for now too if `modulepreload` doesn't work.

Reply:
> Is your understanding of the polyfill that we can't extract the module preload and we should use the whole thing?

I'm not sure about this. But at least this `fetchCache` part cannot be extracted. If this `fetchCache` mechanism is necessary to make the polyfill work, we can't extract only this part.

> The polyfill isn't small and IMO we shouldn't mess with the way modules are imported to support module preload.

I agree to this.

> Then find out if the polyfill is really working.

When I added this comment (https://github.com/vitejs/vite/issues/5532#issuecomment-1100837195), firefox did work when `Cache-Control` was set. I don't know whether it works now though.